doc/developer-guides/modularity.rst

```diff
@@ -0,0 +1,89 @@
+  functionalities in or out. This is crucial in safety-critical scenarios
+  because absence of irrelevant code is required in both MISRA-C and functional
+  safety standards.
+* **Meet functional safety requirements** Functional safety standards explicitly
+  requires a hierarchical structure of modules in software architectural
+  design. This applies to any safety integrity level defined in IEC 61508
+  [IEC 61508-3]_ and ISO 26262 [ISO 26262-6]_.
+
+Principles
+**********
+
+* Each source file shall belong to one module only. One module may consist of
+  one or multiple source files, though. A source file can be a C source, a C
+  header, an assembly file, or any other file that is built into the binary
+  image.
+* Each module shall have well-defined interfaces, including the exported
+  functions and global variables. Functions and variables that are not
+  interfaces shall be static and used inside the module only.
+* Dependencies among the modules should be acyclic. Any cyclic dependency must
+  be deviated explicitly.
+* The complexity of a module shall be limited.
+
+Architecture
+************
+
+The following figure shows the high-level components of ACRN hypervisor. ACRN
+hypervisor adopts a layered design where higher layers can invoke the interfaces
+of lower layers but not vice versa. The only exception is the invocation of
+initialization routine in the **Boot** component.
+
+.. figure:: images/modularity-architecture.png
+   :align: center
+   :name: modularity-architecture
+
+   Layered Architecture of ACRN Hypervisor
+
+The components are listed as follows.
+
+* **Boot** This component carries out the most basic hardware initialization to
+  enable the execution of C code.
+* **Library** This component consists of subroutines that require no explicit
+  initialization. Examples include standard memory and string manipulation
+  functions like strncpy, atomic operations and bitmap operations. This
+  component is independent from and widely used in the other components.
+* **Hardware Management and Utilities** This component abstract hardware
+  resources and provide services like timers and physical interrupt handler
+  registration to the upper layers.
+* **Virtual CPU** This component implements CPU, memory and interrupt
+  virtualization. It handles VM exit events from VMs in general and is able to
+  inject upcall interrupts to a specific VM.
+* **Device Emulation** This component implements devices that are emulated in
+  the hypervisor itself, such as the virtual programmable interrupt controllers
+  including vPIC, vLAPIC and vIOAPIC.
+* **Passthru Management** This component manages devices that are passed-through
+  to specific VMs.
+* **Extended Device Emulation** This component implements an I/O request
+  mechanism that allow the hypervisor to forward I/O accesses from UOSes to SOS
+  for emulation.
+* **VM Management** This component manages the creation, deletion and other
```

Comment:
The box with "vCPU Loop" is not explained.

Reply:
Added. Thanks for pointing this out.